From 07992db236ddbc81da2ce8be58a483c4a88c0e58 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:12:36 -0400
Subject: [powerpc] Clear ELF personality flag if ELFv2 is not requested

Message-id: <1410545655-205645-528-git-send-email-dzickus@redhat.com>
Patchwork-id: 94570
O-Subject: [RHEL7 PATCH 527/626] powerpc: Clear ELF personality flag if ELFv2 is not requested.
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 1efc563865dbc3710c5af7bc1540360a6d06192c
Author: Jeff Bailey <jeffbailey@google.com>
Date:   Sat May 17 15:05:43 2014 +0000

    powerpc: Clear ELF personality flag if ELFv2 is not requested.

    powerpc: Clear ELF personality flag if ELFv2 is not requested.

    The POWER kernel uses a personality flag to determine whether it should
    be setting up function descriptors or not (per the updated ABI).  This
    flag wasn't being cleared on a new process but instead was being
    inherited.  The visible effect was that an ELFv2 binary could not execve
    to an ELFv1 binary.

    Signed-off-by: Jeff Bailey <jeffbailey@google.com>

    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/asm/elf.h b/arch/powerpc/include/asm/elf.h
index 935b5e7..888d8f3 100644
--- a/arch/powerpc/include/asm/elf.h
+++ b/arch/powerpc/include/asm/elf.h
@@ -90,6 +90,8 @@ typedef elf_vrregset_t elf_fpxregset_t;
 do {        \
  if (((ex).e_flags & 0x3) == 2)    \
   set_thread_flag(TIF_ELF2ABI);   \
+ else       \
+  clear_thread_flag(TIF_ELF2ABI);   \
  if ((ex).e_ident[EI_CLASS] == ELFCLASS32)  \
   set_thread_flag(TIF_32BIT);   \
  else       \
-- 
1.7.1